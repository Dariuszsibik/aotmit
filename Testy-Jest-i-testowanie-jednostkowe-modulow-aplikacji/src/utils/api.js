export const login = (email, password) => {
  // w realnej sytuacji mielibyśmy tutaj jakieś zapytanie do api
  // fetch('http://my-login-url.com', {body: JSON.stringify({email,password},),method:'POST'})
  // console.log("Login with:",{email,password})

  return Promise.resolve();
  // return Promise.reject('Niepoprawne dane logowania');
};
