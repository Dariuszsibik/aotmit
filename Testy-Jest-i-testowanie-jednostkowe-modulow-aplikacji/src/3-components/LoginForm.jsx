import { useState } from 'react';
import { isEmail } from '../2-email-validator/validation';
import { login } from '../utils/api.js';

const LoginForm = ({ onSuccess, onFail }) => {
  const [formState, setFormState] = useState({ email: '', password: '' });
  const [error, setError] = useState(null);

  const handleSubmit = (event) => {
    setError(null);
    const { email, password } = formState;
    event.preventDefault();
    
    if (!email || !password) {
      setError('Niepoprawny email lub hasło');
      return;
    }
    if (email && !isEmail(email)) {
      setError('Niepoprawny email');
      return;
    }

    // mock api behavior
    login(email, password)
      .then(() => {
        console.log('login successful');
        onSuccess();
      })
      .catch((error) => {
        console.log('login failed');
        onFail(error);
      });
  };

  const handleFormChange = (event) => {
    const { name, value } = event.target;
    setFormState({ ...formState, [name]: value });
  };

  return (
    <>
      <form onSubmit={handleSubmit}>
        <label>
          Email
          <input id="email" name="email" onChange={handleFormChange} value={formState.email} />
        </label>
        <label>
          Hasło
          <input id="password" name="password" onChange={handleFormChange} value={formState.password} />
        </label>
        <button id="button-submit" type="submit">Zaloguj się</button>
      </form>
      {error && <p style={{ color: 'red' }}>{error}</p>}
    </>
  );
}

export default LoginForm;
