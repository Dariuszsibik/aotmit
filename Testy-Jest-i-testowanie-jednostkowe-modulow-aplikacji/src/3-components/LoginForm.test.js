import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import LoginForm, { FORM_ERROR } from './LoginForm';
import * as api from '../utils/api';

describe('LoginForm', () => {
  it('should render successfully', () => {
      const { baseElement, debug } = render(<LoginForm />);

      const email = baseElement.querySelector('#email');
      const password = baseElement.querySelector('#password');
      const button = baseElement.querySelector('#button-submit');

      expect(email).toBeVisible();
      expect(password).toBeVisible();
      expect(button).toBeVisible();
      expect(baseElement).toBeTruthy();

  });
  it('should display error when email is invalid', () => {
    const { baseElement, debug } = render(<LoginForm />);

    const email = baseElement.querySelector('#email');
    const password = baseElement.querySelector('#password');
    const button = baseElement.querySelector('#button-submit');

    fireEvent.change(email, { target: { value: 'email' } });
    fireEvent.change(password, { target: { value: '12345' } });
    fireEvent.click(button);

    expect(screen.getByText('Niepoprawny email')).toBeVisible();
  });
  it('should display error when missing email fields data', () => {
    const { baseElement, debug } = render(<LoginForm />);

    const email = baseElement.querySelector('#email');
    const password = baseElement.querySelector('#password');
    const button = baseElement.querySelector('#button-submit');

    fireEvent.change(password, { target: { value: '12345' } });
    fireEvent.click(button);

    expect(screen.getByText('Niepoprawny email lub hasło')).toBeVisible();
  });
  it('should display error when missing password fields data', () => {
    const { baseElement, debug } = render(<LoginForm />);

    const email = baseElement.querySelector('#email');
    const password = baseElement.querySelector('#password');
    const button = baseElement.querySelector('#button-submit');

    fireEvent.change(email, { target: { value: 'email' } });
    fireEvent.click(button);

    expect(screen.getByText('Niepoprawny email lub hasło')).toBeVisible();
  });

  // DRUGA CZĘŚĆ ZADANIA - MOCKI
  it('should call onSuccess when login is accepted', async () => {
        // given
        const onSuccessSpy = jest.fn();
        const onFailedSpy = jest.fn();
        const { baseElement, debug } = render(<LoginForm onSuccess={onSuccessSpy} onFail={onFailedSpy} />);

        const email = baseElement.querySelector('#email');
        const password = baseElement.querySelector('#password');
        const buttonSubmit = baseElement.querySelector('#button-submit');
        const loginSpy = jest.spyOn(api, 'login').mockRejectedValue('My custom error message');

        // when
        fireEvent.change(email, { target: { value: 'mariuszkowalski@gmail.com' } } );
        fireEvent.change(password, { target: { value: 'test123' } } );
        fireEvent.click(buttonSubmit);

        await waitFor(() => {
            expect(onFailedSpy).toHaveBeenCalled();
            expect(onSuccessSpy).not.toHaveBeenCalled();
        })

        // then
        expect(loginSpy).toHaveBeenCalled();
  });
  it.todo('should call onFail when login is rejected');

  // it('snapshot', () => {
  //   const { baseElement } = render(<LoginForm />);
  //   expect(baseElement).toMatchSnapshot();
  // });
});
