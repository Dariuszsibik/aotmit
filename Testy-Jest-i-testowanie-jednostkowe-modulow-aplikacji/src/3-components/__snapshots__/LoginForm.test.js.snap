// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm snapshot 1`] = `
<body>
  <div>
    <form>
      <label>
        Email
        <input
          id="email"
          name="email"
          value=""
        />
      </label>
      <label>
        Hasło
        <input
          id="password"
          name="password"
          value=""
        />
      </label>
      <button
        id="button-submit"
        type="submit"
      >
        Zaloguj się
      </button>
    </form>
  </div>
</body>
`;
